ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java

```diff
@@ -0,0 +1,171 @@
+
+public class WebSocketModule extends ReactContextBaseJavaModule {
+
+  private Map<Integer, WebSocket> mWebSocketConnections = new HashMap<>();
+  private ReactContext mReactContext;
+
+  public WebSocketModule(ReactApplicationContext ctx) {
+    super(ctx);
+    mReactContext = ctx;
+  }
+
+  private void sendEvent(String eventName, @Nullable WritableMap params) {
+    mReactContext
+      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
+      .emit(eventName, params);
+  }
+
+  @Override
+  public String getName() {
+    return "WebSocketModule";
+  }
+
+  @ReactMethod
+  public void connect(final String url, final int id) {
+    OkHttpClient client = new OkHttpClient();
+
+    client.setConnectTimeout(10, TimeUnit.SECONDS);
+    client.setWriteTimeout(10, TimeUnit.SECONDS);
+    // Disable timeouts for read
+    client.setReadTimeout(0, TimeUnit.MINUTES);
+
+    Request request = new Request.Builder()
+        .tag(id)
+        .url(url)
+        .build();
+
+    WebSocketCall.create(client, request).enqueue(new WebSocketListener() {
+      @Override
+      public void onOpen(final WebSocket webSocket, final Response response) {
+        mWebSocketConnections.put(id, webSocket);
+
+        WritableMap params = Arguments.createMap();
+        params.putInt("id", id);
+        sendEvent("websocketOpen", params);
+      }
+
+      @Override
+      public void onMessage(
+          final BufferedSource bufferedSource,
+          final WebSocket.PayloadType payloadType
+      ) {
+        String message = "";
+
+        try {
+          message = bufferedSource.readUtf8();
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+
+        WritableMap params = Arguments.createMap();
+
+        params.putInt("id", id);
+        params.putString("data", message);
+
+        try {
+          bufferedSource.close();
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+
+        sendEvent("websocketMessage", params);
+      }
+
+      @Override
+      public void onClose(final int code, final String reason) {
+        WritableMap params = Arguments.createMap();
+
+        params.putInt("id", id);
+        params.putInt("code", code);
+        params.putString("reason", reason);
+
+        sendEvent("websocketClosed", params);
+      }
+
+      @Override
+      public void onFailure(final IOException e, final Response response) {
+        WritableMap params = Arguments.createMap();
+
+        params.putInt("id", id);
+        params.putString("message", e.getMessage());
+
+        sendEvent("websocketFailed", params);
+      }
+
+      @Override
+      public void onPong(final Buffer buffer) {
+      }
+    });
+
+    client.getDispatcher().getExecutorService().shutdown();
+  }
+
+  @ReactMethod
+  public void close(final int code, final String reason, final int id) {
+    WebSocket mWebSocketClient = mWebSocketConnections.get(id);
+
+    if (mWebSocketClient != null) {
```

Comment:
To reduce indentation, you can just do:

if (mWebSocketClient != null) {
    return;
}

Maybe `Flog.w` or even throw an exception if that happens? Why would it be null? Does it mean the developer is passing a wrong id from JS? In that case we should probably redbox/crash the app because it's a developer error that should "never happen".